src/java.base/share/classes/java/lang/ClassValue.java

```diff
@@ -220,29 +212,52 @@ private T getFromBackup(Entry<?>[] cache, Class<?> type) {
     private T getFromHashMap(Class<?> type) {
         // The fail-safe recovery is to fall back to the underlying classValueMap.
         ClassValueMap map = getMap(type);
-        for (;;) {
-            Entry<T> e = map.startEntry(this);
-            if (!e.isPromise())
-                return e.value();
+        var accessed = map.readAccess(this);
+        if (accessed instanceof Entry) {
+            @SuppressWarnings("unchecked")
+            var cast = (Entry<T>) accessed;
+            return cast.value;
+        }
+
+        RemovalToken permission = (RemovalToken) accessed; // nullable
+        for (; ; ) {
+            T value;
             try {
-                // Try to make a real entry for the promised version.
-                e = makeEntry(e.version(), computeValue(type));
-            } finally {
-                // Whether computeValue throws or returns normally,
-                // be sure to remove the empty entry.
-                e = map.finishEntry(this, e);
+                value = computeValue(type);
+            } catch (Throwable ex) {
+                // no value is associated, but there may be already associated
+                // value. Return that if it exists.
+                accessed = map.readAccess(this);
+                if (accessed instanceof Entry) {
+                    @SuppressWarnings("unchecked")
+                    var cast = (Entry<T>) accessed;
+                    return cast.value;
+                }
+                // report failure here, but allow other callers to try again
+                if (ex instanceof RuntimeException rte) {
+                    throw rte;
+                } else {
+                    throw ex instanceof Error err ? err : new Error(ex);
+                }
```

Comment:
Hello Chen, i think this if/else for exception checks can instead just be replaced with an unconditional `throw ex`;

Reply:
I tried that and failed compilation before

Reply:
This could also use `jdk.internal.misc.Unsafe​::throwException​(Throwable)`:
```java
Unsafe.getUnsafe().throwException(ex);
return null;
```